keps/sig-windows/2258-node-service-log-viewer/README.md

```diff
@@ -0,0 +1,395 @@
+
+If a kubectl version that has the new `logs nodes --service` option is used
+against a node that is using a kubelet that does not have the extended
+`/var/log` endpoint viewer, the result should be "feature not supported".
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate
+    - Feature gate name: NodeLogs
+    - Components depending on the feature gate: kubelet
+  - [x] Other
+    - Describe the mechanism:
+    - Will enabling / disabling the feature require downtime of the control
+      plane? No
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? It will require downtime per node.
+
+* **Does enabling the feature change any default behavior?** No
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?** Yes. It can be disabled by disabling the `NodeLogs` feature
+  gate in the kubelet.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+  There will be no adverse effects of enabling the feature gate after it was
+  disabled.
+
+* **Are there any tests for feature enablement/disablement?** No
+
+### Rollout, Upgrade and Rollback Planning
+
+_This section must be completed when targeting beta graduation to a release._
+
+### Monitoring Requirements
+
+_This section must be completed when targeting beta graduation to a release._
+
+### Dependencies
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **Does this feature depend on any specific services running in the cluster?**
+  - kubelet
+    - Usage description:
+      - Impact of its outage on the feature: If kubelet is not running on the
+        this feature will not work.
+      - Impact of its degraded performance or high-error rates on the feature:
+        If the kubelet is degraded this feature will also be degraded i.e. the
+        node logs will not be returned.
+
+### Scalability
+
+* **Will enabling / using this feature result in any new API calls?**
+  No
+
+* **Will enabling / using this feature result in introducing new API types?**
+  Yes. We will need to add a `NodeLogOptions` counterpart to
+  [PodLogOptions](https://github.com/kubernetes/kubernetes/blob/548ad1b8d35d51e6d33ea21dcc75d60a789b00e6/pkg/apis/core/types.go#L4409)
+  
+* **Will enabling / using this feature result in any new calls to the cloud 
+provider?**
+  No
+
+* **Will enabling / using this feature result in increasing size or count of 
+the existing API objects?**
+  No
+
+* **Will enabling / using this feature result in increasing time taken by any 
+operations covered by [existing SLIs/SLOs]?**
+  No
+
+* **Will enabling / using this feature result in non-negligible increase of 
+resource usage (CPU, RAM, disk, IO, ...) in any components?**
+  In the case of large logs, there is potential for an increase in RAM and CPU
```

Comment:
could this be used to DOS the kubelet?

Reply:
That's possibility and we plan to collect feedback about it during the alpha. This is also why it has been restricted to cluster admins. Please see https://github.com/kubernetes/enhancements/pull/2271/commits/2450f181852d96378815014c0ff6186d512818ce#r568930284

Reply:
In practice reading journal via this endpoint will have a slightly higher CPU and IO cost than reading a file directly off the kubelet.  However, the total cost of the journalctl read should be on par or slightly less than pod log reading because the journal has more indexing and also a more aggressive set of optimizations around data reading than pod logs.  That said, there is more complexity inherent in the query mechanism of journalctl and so it is possible that an attacker could generate more load on the system than necessary due to bugs in the journal behavior.  Confined to admins this feature should be roughly proprotionally as expensive as any other admin command, and an admin has other ways to attack the node.  

I would be against exposing this to admins who don't own the node (i.e. cluster-admins or a set of users who have node-admin roles) and that was the same decision we made putting this into openshift.

Reply:
>I would be against exposing this to admins who don't own the node (i.e. cluster-admins or a set of users who have node-admin roles) and that was the same decision we made putting this into openshift.

We discussed this with `sig-cli` and @soltysh seemed to think this might not be doable with `kubectl`. @soltysh please weigh in here as I could have misunderstood what you were describing.

Reply:
I just realized `not` had slipped in my previous comment. I thought we are indeed restricting this to cluster admins who have the node admin role given that the logs command constructs and sends an API call and if the objects referenced in the API call can't be accessed by the user sending it because there is an RBAC policy saying so, the API server will return an error. Or are you saying we should explicitly check if the user has the node-admin role in the `kubectl` code?

Reply:
i'm saying the logs API call needs no change, and we shouldn't advocate adding more people to /logs without a different discussion (this KEP is about keeping the same security profile).  It's always ok to expose things to kubectl that some users can't do (some users can't delete cluster roles, doesn't mean kubectl limits the delete command).